tests/test_regularization.py

```diff
@@ -0,0 +1,436 @@
+        th.tensor(0.1),
+        th.tensor(0.0),
+        100.0,
+        10.0,
+        0.1,
+        0.0,
+    ],
+)
+def test_interval_param_scaler(lambda_, train_loss, interval_param_scaler):
+    scaler = interval_param_scaler
+    tolerable_interval = scaler.tolerable_interval
+    scaling_factor = scaler.scaling_factor
+    eps = np.finfo(float).eps
+    if train_loss > eps:
+        # The loss is a non-zero scalar, so we can construct a validation loss for
+        # three different cases:
+
+        # case that the ratio between the validation loss and the training loss is
+        # above the tolerable interval
+        val_loss = train_loss * tolerable_interval[1] * 2
+        assert scaler(lambda_, train_loss, val_loss) == lambda_ * (1 + scaling_factor)
+
+        # case that the ratio between the validation loss and the training loss is
+        # below the tolerable interval
+        val_loss = train_loss * tolerable_interval[0] / 2
+        assert scaler(lambda_, train_loss, val_loss) == lambda_ * (1 - scaling_factor)
+
+        # case that the ratio between the validation loss and the training loss is
+        # within the tolerable interval
+        val_loss = train_loss * (tolerable_interval[0] + tolerable_interval[1]) / 2
+        assert scaler(lambda_, train_loss, val_loss) == lambda_
+    else:
+        # we have a zero loss. We try two cases. When the validation loss is zero,
+        # the ratio is undefined, so we should return the current lambda. When the
+        # validation loss is nonzero, the ratio is infinite, so we should see the lambda
+        # increase by the scaling factor.
+        # We try it for both a tensor and a float value.
+        val_loss = th.tensor(0.0)
+        assert scaler(lambda_, train_loss, val_loss) == lambda_
+        val_loss = 0.0
+        assert scaler(lambda_, train_loss, val_loss) == lambda_
+        val_loss = th.tensor(1.0)
+        assert scaler(lambda_, train_loss, val_loss) == lambda_ * (1 + scaling_factor)
+        val_loss = 1.0
+        assert scaler(lambda_, train_loss, val_loss) == lambda_ * (1 + scaling_factor)
+
+
+def test_interval_param_scaler_raises(interval_param_scaler):
+    scaler = interval_param_scaler
+    with pytest.raises(ValueError, match="val_loss and train_loss must be scalars"):
+        scaler(0.0, th.Tensor([1.0, 2.0]), th.Tensor([3.0, 4.0]))
+    with pytest.raises(ValueError, match="val_loss and train_loss must be scalars"):
+        scaler(0.0, "random value", "random value")  # type: ignore
+    with pytest.raises(ValueError, match="lambda_ must be a float"):
+        scaler(th.tensor(1.0), 1.0, 1.0)  # type: ignore
+    with pytest.raises(ValueError, match="lambda_ must not be zero.*"):
+        scaler(0.0, 1.0, 1.0)
+
+
+def test_interval_param_scaler_init_raises():
+    # this validates the value of scaling_factor.
+    with pytest.raises(
+        ValueError,
+        match=r"scaling_factor must be in \(0, 1\) within machine precision.",
```

Comment:
Could assign this to a variable and reuse it rather than having to copy-paste it.

Heck you could even assign the whole `pytest.raises(...)` to a context manager though that's maybe sufficiently counterintuitive to be not worth the lines saved.

Reply:
I could write a separate test and parametrize it, but I think commenting why each set of values raises is useful enough to leave it this way.